default.nix

```diff
@@ -118,30 +125,46 @@ with stdenv.lib; let
 in rec {
   mkNodeModules = { src, extraEnvVars ? {} }:
     let
+      localDeps = map dirOfLocal (builtins.filter isLocal (builtins.attrValues (lock.dependencies)));
+
+      origSrc = if src ? _isLibCleanSourceWith then src.origSrc else src;
+      getRelativePath = path: removePrefix (toString origSrc + "/") path;
+
       # filter out everything except package.json and package-lock.json if possible
       # allows to avoid rebuilding node_modules if these two files didn't change
       filteredSrc =
         let
-          origSrc = if src ? _isLibCleanSourceWith then src.origSrc else src;
-          getRelativePath = path: removePrefix (toString origSrc + "/") path;
-          usedPaths = [ "package.json" "package-lock.json" ];
+          usedPaths = [ "package.json" "package-lock.json" ]
+            ++ localDeps;
+          dirOfRec = x: if dirOf x == "." || dirOf x == "/" then [x] else ([x] ++ dirOfRec (dirOf x));
+          usedPathsRec = builtins.concatMap dirOfRec usedPaths;
           cleanedSource = cleanSourceWith {
             src = src;
-            filter = path: type: elem (getRelativePath path) usedPaths;
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
           };
         in if canCleanSource src then cleanedSource else src;
 
-      packageJson = filteredSrc + "/package.json";
-      packageLockJson = filteredSrc + "/package-lock.json";
+      metaSrc = cleanSourceWith {
+        src = src;
+        filter = path: type: elem (getRelativePath path) [ "package.json" "package-lock.json" ];
+      };
```

Comment:
`filteredSrc` also has a check for `canCleanSource` to support the situation when `src` is a nix store path not produced by `cleanSourceWith`, it should be done here as well. Also, to avoid repetition, you can make `filteredSrc` a function which accepts `localDeps` as a parameter and use `filteredSrc []` for `metaSrc`